### PR TITLE
Release 0.31.0

### PR DESCRIPTION
**Pull request by bcardiff**
Preview at https://github.com/bcardiff/crystal/blob/changelog/0.31.0/CHANGELOG.md

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,114 @@
...
+### Collections
+
+- **(performance)** Array optimizations for small number of elements. ([#8048](https://github.com/crystal-lang/crystal/pull/8048), thanks @asterite)
+- **(performance)** Optimize `Array#*`. ([#8087](https://github.com/crystal-lang/crystal/pull/8087), thanks @asterite)
+- **(performance)** Hash now uses an open addressing algorithm. ([#8017](https://github.com/crystal-lang/crystal/pull/8017), [#8182](https://github.com/crystal-lang/crystal/pull/8182), thanks @asterite)
+- **(performance)** Optimize `Hash#to_a`, `Hash#keys` and `Hash#values`. ([#8042](https://github.com/crystal-lang/crystal/pull/8042), thanks @asterite)
+- **(performance)** Add `Hash#put` and optimize `Set#add?`. ([#8116](https://github.com/crystal-lang/crystal/pull/8116), thanks @asterite)
+- Fixed `Slice#==` for some generic instantiations, add `Slice#<=>`. ([#8074](https://github.com/crystal-lang/crystal/pull/8074), thanks @asterite)
+- Add docs on idempotence and methods involving eager evaluation in `Iterator`. ([#8053](https://github.com/crystal-lang/crystal/pull/8053), thanks @KimBurgess)
+- Add `Set#|`. ([#8121](https://github.com/crystal-lang/crystal/pull/8121), thanks @sam0x17)
```
Comment (Sija) [inline CHANGELOG.md]:
```suggestion
- Add `Set#+`. ([#8121](https://github.com/crystal-lang/crystal/pull/8121), thanks @sam0x17)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,114 @@
...
+
+- **(breaking-change)** Move `Markdown` to inside the compiler. ([#8115](https://github.com/crystal-lang/crystal/pull/8115), thanks @asterite)
+- **(breaking-change)** Deprecate `OptionParser#parse!`, use `OptionParser#parse`. ([#8041](https://github.com/crystal-lang/crystal/pull/8041), thanks @didactic-drunk)
+
+### Numeric
+
+- **(breaking-change)** Enable overflow by default. ([#8170](https://github.com/crystal-lang/crystal/pull/8170), thanks @bcardiff)
+- **(breaking-change)** Make `/` the arithmetic division for all types. ([#8120](https://github.com/crystal-lang/crystal/pull/8120), thanks @bcardiff)
+- Add `BigDecimal#**` and `BigRational#**` (pow operator). ([#7860](https://github.com/crystal-lang/crystal/pull/7860), thanks @jwbuiter)
+
```
Comment (straight-shoota) [inline CHANGELOG.md]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,119 @@
+# 0.31.0 (2019-09-??)
+
+## Language changes
+
+- Allow anonymous block args and their type restriction `& : T -> U`. ([#8117](https://github.com/crystal-lang/crystal/pull/8117), thanks @asterite)
```
Comment (RX14) [inline CHANGELOG.md]: Allow specifying block args without a name?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,119 @@
...
+
+### Macros
+
+- Ensure `@type` is devirtualized inside macros. ([#8149](https://github.com/crystal-lang/crystal/pull/8149), thanks @asterite)
+
+## Standard library
+
+- **(breaking-change)** Move `Markdown` to inside the compiler. ([#8115](https://github.com/crystal-lang/crystal/pull/8115), thanks @asterite)
```
Comment (RX14) [inline CHANGELOG.md]: I'd frame this as removing it from the stdlib, not moving it into the compiler (thats an implementation detail)

File: CHANGELOG.md (hunk 3 above)
Comment (bcardiff) [reply CHANGELOG.md]: nameless block args? I was looking for a descriptive name as a whole

Comment (straight-shoota) [reply CHANGELOG.md]: The distinctive feature of a nameless block arg is that it is not captured. So I'd refer to this as non-captured block. This term is also used in the language reference.

**Pull request merged**